### PR TITLE
Fix exception propagation in Async API methods

### PR DESCRIPTION
**Pull request by vbabanin**
- Resolve an issue where exceptions thrown during thenRun, thenSupply, and related operations in the asynchronous API were not properly propagated to the completion callback. This issue was addressed by replacing `unsafeFinish` with `finish`, ensuring that exceptions are caught and correctly passed to the completion callback when executed on different threads.

- Update existing Async API tests to ensure they simulate separate async thread execution.

JAVA-5562

### CODE REVIEW

**Review by rozza — CHANGES_REQUESTED** [review]
Looks really good.

However, one test in evergreen seems to be failing: `com.mongodb.reactivestreams.client.unified.LoadBalancerTest` which needs investigation before I can LGTM

File: driver-core/src/main/com/mongodb/internal/async/AsyncFunction.java [hunk 1]
```diff
@@ -33,4 +35,28 @@ public interface AsyncFunction<T, R> {
...
+     * Must be invoked at end of async chain or when executing a callback handler supplied by the caller.
+     *
+     * @param callback the callback provided by the method the chain is used in.
+     */
+    default void finish(final T value, final SingleResultCallback<R> callback) {
+        final AtomicBoolean callbackInvoked = new AtomicBoolean(false);
+        try {
+            this.unsafeFinish(value, (v, e) -> {
+                if (!callbackInvoked.compareAndSet(false, true)) {
+                    throw new AssertionError("Callback has been already completed. It could happen "
```
[the remark this answers is not in the document]
Comment (vbabanin) [reply driver-core/src/main/com/mongodb/internal/async/AsyncFunction.java]: Makes sense! I have included them in the message. Thanks!

**Review by rozza — APPROVED** [review]
LGTM!

**Comment by vbabanin** [comment]
The Async API is currently utilized in the OIDC, Authenticator, and InternalStreamConnection components. I have conducted additional OIDC tests to verify the changes, and all results are green::

[OIDC Auth Test](https://spruce.mongodb.com/task/mongo_java_driver_oidc_auth_test_oidc_auth_test_patch_81402ae070244dfe5ba8f70d93b00798d8dd39c7_66bfeb9414f2af00076e9ceb_24_08_17_00_15_18/logs?execution=0)
[Azure OIDC Auth Test](https://spruce.mongodb.com/task/mongo_java_driver_testazureoidc_variant_oidc_auth_test_azure_patch_81402ae070244dfe5ba8f70d93b00798d8dd39c7_66bfeb9414f2af00076e9ceb_24_08_17_00_15_18/logs?execution=1)
[GCP OIDC Auth Test](https://spruce.mongodb.com/task/mongo_java_driver_testgcpoidc_variant_oidc_auth_test_gcp_patch_81402ae070244dfe5ba8f70d93b00798d8dd39c7_66bfeb9414f2af00076e9ceb_24_08_17_00_15_18/logs?execution=0)

**Pull request merged**